### ISSUE TITLE
beta1: test_cmp_cli not honouring the EXE_SHELL env variable

### ISSUE THREAD

**Issue by dlegaultbbry** [openssl/openssl#15833]
I use a non default method of testing when porting openssl to QNX in a cross compiling environment and noted that this test doesn't work correctly with EXE_SHELL defined. All other tests are successful with this override.

EXE_SHELL="perl $(pwd)/util/wrap.pl" perl test/run_tests.pl

Compared to other tests it seems to use run(cmd()) instead of run(test()) and cmd doesn't receive the EXE_SHELL option when called.

```
sh: ../../util/wrap.pl: No such file or directory
../../util/wrap.pl ../../apps/openssl cmp -config "" -help => 1
not ok 1 - show help
===(    2723;2639  1110/1144   8/12  2/6  7/9 )=========================--------
sh: ../../util/wrap.pl: No such file or directory
../../util/wrap.pl ../../apps/openssl cmp -config "" -poll_count 1 -use_mock_srv -srv_ref 'mock server' -srv_secret 'pass:test' -rsp_cert ../../test/certs/ee-cert-1024.pem -cmd cr -subject /CN=any -newkey ../../test/certs/ee-key-1024.pem -secret 'pass:test' -ref 'client under test' -certout ./test.certout.pem => 1
not ok 8 - with polling
# ------------------------------------------------------------------------------
#   Failed test 'with polling'
#   at test/recipes/81-test_cmp_cli.t line 68.
sh: ../../util/wrap.pl: No such file or directory
../../util/wrap.pl ../../apps/openssl cmp -config "" -engine loader_attic -use_mock_srv -srv_ref 'mock server' -srv_secret 'pass:test' -rsp_cert ../../test/certs/ee-cert-1024.pem -cmd cr -subject /CN=any -newkey ../../test/certs/ee-key-1024.pem -secret 'pass:test' -ref 'client under test' -certout ./test.certout.pem => 1
not ok 9 - with loader_attic engine
# ------------------------------------------------------------------------------
#   Failed test 'with loader_attic engine'
#   at test/recipes/81-test_cmp_cli.t line 68.
81-test_cmp_cli.t .................. Dubious, test returned 3 (wstat 768, 0x300)
Failed 3/9 subtests 
```

**Comment by levitte** [comment]
Please try #15846

**Comment by dlegaultbbry** [comment]
But it's not the failing test that I reported. This one doesn't run in cross compile settings:

`80-test_cmp_http.t ................. skipped: Tests involving local HTTP server not available in cross-compile builds`

This is the test that is failing for me:

`81-test_cmp_cli.t .................. Dubious, test returned 3 (wstat 768, 0x300)`

**Comment by levitte** [comment]
Ah, oops

**Comment by levitte** [comment]
@dlegaultbbry, look now

**Comment by dlegaultbbry** [comment]
Latest patch works, but I forgot to mention that the below fix is also required to make things work nicely with the EXE_SHELL override for some other tests. This is my local fix, maybe there's another way to fix it but I'm not a perl expert.

The first sub is used in 80-test_ca, the second is used in 80-test_ssl_new

```
diff --git a/lib/openssl3/dist/util/perl/OpenSSL/Test.pm b/lib/openssl3/dist/util/perl/OpenSSL/Test.pm
index 00aa4d841e..30a4f94dd2 100644
--- a/lib/openssl3/dist/util/perl/OpenSSL/Test.pm
+++ b/lib/openssl3/dist/util/perl/OpenSSL/Test.pm
@@ -363,7 +363,7 @@ sub perlapp {
         my @cmdargs = ( @{$cmd} );
         my @prog = __apps_file(shift @cmdargs, undef);
         return cmd([ @interpreter, @interpreter_args,
-                     @prog, @cmdargs ], %opts) -> (shift);
+                     @prog, @cmdargs ], exe_shell => $ENV{EXE_SHELL}, %opts) -> (shift);
     }
 }
 
@@ -377,7 +377,7 @@ sub perltest {
         my @cmdargs = ( @{$cmd} );
         my @prog = __test_file(shift @cmdargs, undef);
         return cmd([ @interpreter, @interpreter_args,
-                     @prog, @cmdargs ], %opts) -> (shift);
+                     @prog, @cmdargs ], exe_shell => $ENV{EXE_SHELL}, %opts) -> (shift);
     }
 }
```

**Comment by levitte** [comment]
Patch perlapp() and perltest() like that is pretty meaningless.

EXE_SHELL is designed for use in cross-compiled builds, where you want to run with the build host perl.  An example is to cross-compile for mingw on Linux, and then run the tests with `EXE_SHELL=wine`.

Speaking of which, I have no idea why you have `EXE_SHELL="perl $(pwd)/util/wrap.pl"`.  Have you tried without it?

**Comment by dlegaultbbry** [comment]
I can't test with the host perl since I'm cross-compiling for QNX targets. The way we have this setup is that we just cross-compile all the openssl libraries and tests and then just copy the whole source tree in place on a QNX target device and run the tests locally as you would when compiling for darwin/linux. We need to fiddle slightly with expected paths of things which are not the same as the host since openssl now hardcodes paths from the host in cross-compilation results. I did not have to do any of this in 1.1.1 so this is a change of how things worked previously.

I'm fine if you don't want to integrate this fix, I'll just keep it locally so that things just work in our setup. This is the quick way I found to get these tests running because I need to override the whole env with EXE_SHELL to point to the location of the wrap.pl script.
